File: eyedb-bench/trunk/src/java/org/eyedb/benchmark/framework/StopWatch.java
package org.eyedb.benchmark.framework;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author Fran&ccedil;ois D&eacute;chelle (francois@dechelle.net)
 */

public class StopWatch {

    public StopWatch()
    {
	running = false;
	startTime = totalTime = lapTime = 0L;
	laps = new ArrayList<Map.Entry<String,Long>>();
    }

    public final void start()
    {
	running = true;
	startTime = systemTime();
	lapTime = 0;
    }

    public long stop()
    {
	if (running) {
	    running = false;
	    totalTime = time();
	} else 
	    totalTime = 0;

	return totalTime;
    }

    public long lap( String name)
    {
	long t = time();
	long delta = t - lapTime;

	Map.Entry e = new AbstractMap.SimpleEntry<String,Long>( name, new Long(delta));

	laps.add( e);
	lapTime = t;

	return delta;
    }

    public void reset()
    {
	running = false;
	startTime = totalTime = lapTime = 0L;
	laps.clear();
    }

    public long getTotalTime()
    {
	return totalTime;
    }

    public List< Map.Entry<String,Long> > getLaps()
    {
	return laps;
    }

    private long systemTime()
    {
	return System.currentTimeMillis();
    }

    private long time()
    {
	return systemTime() - startTime;
    }

    private boolean running;
    private long startTime;
    private long totalTime;
    private long lapTime;
    private List< Map.Entry<String,Long> > laps;
}

File: eyedb-bench/trunk/src/cpp/benchmark/stopwatch.cc
#include "stopwatch.h"
#include <sys/time.h>
#include <time.h>
#include <iostream>

using namespace eyedb::benchmark;

void StopWatch::start()
{
  running = true;
  startTime = systemTime();
  lapTime = 0;
}

unsigned long StopWatch::stop()
{
  if (running) {
    running = false;
    totalTime = time();
  } else
    totalTime = 0;

  return totalTime;
}

unsigned long StopWatch::lap( const std::string &name)
{
  unsigned long t = time();
  unsigned long delta = t - lapTime;
  laps.push_back( Lap( name, delta));
  lapTime = t;

  return delta;
}

void StopWatch::reset()
{
  running = false; 
  startTime = totalTime = lapTime = 0;
  laps.clear();
}

unsigned long StopWatch::systemTime()
{
  struct timeval now;

  gettimeofday( &now, NULL);

  return 1000 * now.tv_sec + now.tv_usec / 1000;
}
